DuckDuckGo/Freemium/PIR/FreemiumPIRFeature.swift

```diff
@@ -0,0 +1,77 @@
+//
+//  FreemiumPIRFeature.swift
+//
+//  Copyright © 2024 DuckDuckGo. All rights reserved.
+//
+//  Licensed under the Apache License, Version 2.0 (the "License");
+//  you may not use this file except in compliance with the License.
+//  You may obtain a copy of the License at
+//
+//  http://www.apache.org/licenses/LICENSE-2.0
+//
+//  Unless required by applicable law or agreed to in writing, software
+//  distributed under the License is distributed on an "AS IS" BASIS,
+//  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//  See the License for the specific language governing permissions and
+//  limitations under the License.
+//
+
+import Foundation
+import BrowserServicesKit
+import Subscription
+import Freemium
+
+/// Conforming types encapsulate logic relating to the Freemium PIR Feature (e.g Feature Availability etc.)
+protocol FreemiumPIRFeature {
+    var isAvailable: Bool { get }
+}
+
+/// Default implementation of `FreemiumPIRFeature`
+final class DefaultFreemiumPIRFeature: FreemiumPIRFeature {
+
+    private let featureFlagger: FeatureFlagger
+    private let subscriptionManager: SubscriptionManager
+    private let accountManager: AccountManager
+
+    var isAvailable: Bool {
+        /* Freemium PIR availability criteria:
+            1. Feature Flag enabled
+            2. Privacy Pro Available
+            3. Not a current Privacy Pro subscriber
+            4. (Temp) In experiment cohort
+         */
+        featureFlagger.isFeatureOn(.freemiumPIR) // #1
+        && subscriptionManager.isPrivacyProAvailable // #2
+        && !accountManager.hasSubscription // #3
+        // TODO: - Also check experiment cohort here
+    }
+
+    init(featureFlagger: FeatureFlagger = NSApp.delegateTyped.featureFlagger,
+         subscriptionManager: SubscriptionManager,
+         accountManager: AccountManager) {
+
+        self.featureFlagger = featureFlagger
+        self.subscriptionManager = subscriptionManager
+        self.accountManager = accountManager
+    }
+}
+
+private extension SubscriptionManager {
+
+    var isPrivacyProAvailable: Bool {
+        let platform = currentEnvironment.purchasePlatform
+        switch platform {
+        case .appStore:
+            return canPurchase
+        case .stripe:
+            return true
+        }
+    }
+}
+
+private extension AccountManager {
+
+    var hasSubscription: Bool {
```

Comment:
What is the point of renaming an already used property across the code base?

Reply:
This is just creating a new, IMO more descriptive property which is local (private) to this type. This makes it clearer what is being checked at the call site, e.g:

```
featureFlagger.isFeatureOn(.freemiumPIR) // #1
        && subscriptionManager.isPrivacyProPurchaseAvailable // #2
        && !accountManager.hasSubscription
```

I think this is better because `isUserAuthenticated` does not automatically equate to `hasSubscription` (i.e some people reading the code might not know that not being authenticated means they don’t have a subscription).

Reply:
I follow your thought process but this further can be problematic in the future (e.g. should `hasSubscription` care about if the subscription is active or not).

My fear is that you are adding a new meaning to already existing state variable. And as it was already mentioned around TD work, from the BE point of view having account (to which we are being authenticated) and having a subscription are two totally separate concepts. For the time being for mobile clients it is true that being authenticated equals having a matching subscription on the BE and this subscription can be either active or expired.

Reply:
Good point re: "from the BE point of view having account (to which we are being authenticated) and having a subscription are two totally separate concepts”. Will remove it based on this.